### ISSUE TITLE
Combining Static Scoping and Type Inference 

### ISSUE THREAD

**Issue by dannyech** [alexandramedway/odds#57]
This is very hard to explain over text so please bear with me. Basically, we need to come up with a way to add constraints to variable types in the scope levels above the current scope.

i.e.

say we have var `a` which is currently unconstrained. Then we have
```ocaml 
  do add = () -> return a + 5
```
We need to constrain `a` while analyzing the add function in this case.

In this case, however, adding constraints would be a mistake:
```ocaml
  do foo = () -> do a = "Hi" return "world"
```

TLDR: when adding constraints we need to match against both regular name and statically scoped name. This is difficult to do (and highly inefficient) with the current way we are passing around the environment. @akalicki and I came up with a more efficient/better way of doing it, however this require a bit of refactoring of code we already thought we had worked out.

**Comment by dannyech** [comment]
Proposal for new environment:
```ocaml
type environment = {
  
scope: Sast.var StringMap.t;
  parent: environment;
}
```
and
```ocaml

**Comment by dannyech** [comment]
Issue: Each variable lookup now has time complexity O(n) where n is the number of scopes down we are. This is not the best solution.

**Comment by akalicki** [comment]
This issue overlaps a bunch with #63 so I'm closing it for now.
